pkg/plugins/manager/manager.go

```diff
@@ -2,268 +2,249 @@ package manager
+			return plugins.DuplicateError{
+				PluginID:          plugin.ID,
+				ExistingPluginDir: plugin.PluginDir,
+			}
+		}
 
-func (m *PluginManager) Renderer() *plugins.Plugin {
-	for _, p := range m.availablePlugins(context.TODO()) {
-		if p.IsRenderer() {
-			return p
+		if dlOpts.PluginZipURL == "" && dlOpts.Version == "" {
+			return fmt.Errorf("could not determine update options for %s", pluginID)
 		}
-	}
 
-	return nil
-}
+		// remove existing installation of plugin
+		err = m.Remove(ctx, plugin.ID)
+		if err != nil {
+			return err
+		}
 
-func (m *PluginManager) SecretsManager() *plugins.Plugin {
-	for _, p := range m.availablePlugins(context.TODO()) {
-		if p.IsSecretsManager() {
-			return p
+		if dlOpts.PluginZipURL != "" {
+			pluginArchive, err = m.pluginRepo.GetPluginArchiveByURL(ctx, dlOpts.PluginZipURL, compatOpts)
+			if err != nil {
+				return err
+			}
+		} else {
+			pluginArchive, err = m.pluginRepo.GetPluginArchive(ctx, pluginID, dlOpts.Version, compatOpts)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		var err error
+		pluginArchive, err = m.pluginRepo.GetPluginArchive(ctx, pluginID, version, compatOpts)
+		if err != nil {
+			return err
 		}
 	}
 
-	return nil
-}
+	extractedArchive, err := m.pluginStorage.Add(ctx, pluginID, pluginArchive.File)
+	if err != nil {
+		return err
+	}
 
-func (m *PluginManager) Routes() []*plugins.StaticRoute {
-	staticRoutes := make([]*plugins.StaticRoute, 0)
+	// download dependency plugins
+	pathsToScan := []string{extractedArchive.Path}
+	for _, dep := range extractedArchive.Dependencies {
+		m.log.Info("Fetching %s dependencies...", dep.ID)
+		d, err := m.pluginRepo.GetPluginArchive(ctx, dep.ID, dep.Version, compatOpts)
+		if err != nil {
+			return fmt.Errorf("%v: %w", fmt.Sprintf("failed to download plugin %s from repository", dep.ID), err)
+		}
 
-	for _, p := range m.availablePlugins(context.TODO()) {
-		if p.StaticRoute() != nil {
-			staticRoutes = append(staticRoutes, p.StaticRoute())
+		depArchive, err := m.pluginStorage.Add(ctx, dep.ID, d.File)
+		if err != nil {
+			return err
 		}
-	}
-	return staticRoutes
-}
 
-func (m *PluginManager) registerAndStart(ctx context.Context, p *plugins.Plugin) error {
-	if err := m.pluginRegistry.Add(ctx, p); err != nil {
-		return err
+		pathsToScan = append(pathsToScan, depArchive.Path)
 	}
 
-	if !p.IsCorePlugin() {
-		m.log.Info("Plugin registered", "pluginId", p.ID)
+	err = m.loadPlugins(context.Background(), plugins.External, pathsToScan...)
+	if err != nil {
+		m.log.Error("Could not load plugins", "paths", pathsToScan, "err", err)
+		return err
 	}
 
-	return m.start(ctx, p)
+	return nil
 }
 
-func (m *PluginManager) unregisterAndStop(ctx context.Context, p *plugins.Plugin) error {
-	m.log.Debug("Stopping plugin process", "pluginId", p.ID)
-	m.pluginsMu.Lock()
-	defer m.pluginsMu.Unlock()
+func (m *PluginManager) AddFromSource(ctx context.Context, source plugins.PluginSource) error {
```

Comment:
Is this used externally or what's the purpose of this?

Reply:
Ah yes I don't actually need this as part of the interface anymore 😄 Good eye!